### DISCUSSION TITLE
RAG-Playground Discussions!

### DISCUSSION THREAD

**Discussion by Anteemony** [Anteemony/RAG-Playground#51] [Ideas]
Hello @martinoywa @OscarArroyoVega 

The project started with `ConversationalRetrievalChain`
 
During runtime, the Chain creates a standalone question using the chat history and passes the new question to the retriever so that it returns documents relevant to the user’s questions or inferences 

The Chain then passes this new question and context to the LLM. 

The problems with this approach were 
1. The LLM was not aware of the chat history. Questions like “What was my last question” couldn’t be answered. 
2. Sometimes the reformulated question fed to the LLM was very different from the query making the conversation look very less connected and even confusing 

Then to fix this problem, we switched to `RetrievalQA`

and we passed the chat history directly to the LLM answering the question. 

The LLM now has access to the chat history 

But the problem is that the retriever no longer has access to the chat history. So the retrieved documents (context) don’t really match the query. So when this context is passed to the LLM, it assumes that the answers are not available in the document.

**Comment by Anteemony** [comment]
Upon checking the docs, 

There was a simple attribute that could have been passed to `ConversationalRetrievalChain` in order to fix this reformulated question problem mentioned earlier. 

`rephrase_question=False` asks the chain to pass the original question to the LLM instead of  the reformulated one

And the chat history could have been passed directly to the prompt template like was done in the `RetrievalQA` chain

These changes have been implemented and pushed. Find them at [this branch](https://github.com/Anteemony/RAG-Playground/tree/old-chain). Please check it out locally, the chatbot should work without any problems now.

**Reply by Anteemony** [reply]
Now the thing is, this ConversationalRetrievalChain has been deprecated 

Even if we choose to continue with it we should still update our code base sooner or later 

So I guess it should be fine to merge this for now and then go ahead and update the cobase later. Any thoughts on this? @OscarArroyoVega  @martinoywa

**Reply by martinoywa** [reply]
I think we should consider updating the code base first. It doesn't seem to be a significant amount of changes.

**Reply by Anteemony** [reply]
Looking at the docs, I see that the recommended chain handles all of these problems for us.

Yes I agree! It shouldn’t be a problem to update the code base. 

The `max_token_limit` attribute came with ConversationalRetrievalChain. But I should assume the recommended one has something similar

**Reply by martinoywa** [reply]
By scheming through the docs it seems there's no "direct" way to set `max_token_limit` with [create_history_aware_retriever](https://api.python.langchain.com/en/latest/chains/langchain.chains.history_aware_retriever.create_history_aware_retriever.html). But I'll keep exploring it to check for options, although it looks like the old versions give more flexibility.

**Reply by Anteemony** [reply]
You can see from the  `history_aware_retriever` image in the [Add Chat History page](https://python.langchain.com/docs/use_cases/question_answering/chat_history/), that the workflow is exactly what we want to implement.

![image](https://python.langchain.com/assets/images/conversational_retrieval_chain-5c7a96abe29e582bc575a0a0d63f86b0.png)

Yes, the flexibility is what I am concerned about too. After some digging, we should be able to reach a conclusion.

**Reply by Anteemony** [reply]
Hello @martinoywa @OscarArroyoVega

Please take a look #53 when you have the time.

The pull request has the fixes with the updated chain.

And yes! The flexibility remains. 

In fact, it seems that it is even more flexible with the expression format.

Let me know what you think after checking it out.

**Comment by Anteemony** [comment]
Hello @martinoywa @OscarArroyoVega

Please take a look https://github.com/Anteemony/RAG-Playground/pull/53 when you have the time.

The pull request has the fixes with the updated chain.

And yes! The flexibility remains.

In fact, it seems that it is even more flexible with the expression format.

Let me know what you think after checking it out.
